zebra-consensus/src/checkpoint.rs

```diff
@@ -3,315 +3,1098 @@
 //! Checkpoint-based verification uses a list of checkpoint hashes to speed up the
 //! initial chain sync for Zebra. This list is distributed with Zebra.
 //!
-//! The CheckpointVerifier compares each block's `BlockHeaderHash` against the known
-//! checkpoint hashes. If it matches, then the block is verified, and added to the
-//! `ZebraState`. Otherwise, if the block's height is lower than the maximum checkpoint
-//! height, the block awaits the verification of its child block.
+//! The CheckpointVerifier queues pending blocks. Once there is a chain between
+//! the next pair of checkpoints, it verifies all the blocks in that chain.
+//! Verification starts at the first checkpoint, which is the genesis block for the
+//! configured network.
 //!
 //! Verification is provided via a `tower::Service`, to support backpressure and batch
 //! verification.
 
 use futures_util::FutureExt;
 use std::{
-    collections::HashMap,
+    collections::BTreeMap,
     error,
     future::Future,
+    iter::successors,
+    ops::{Bound, Bound::*, RangeBounds},
     pin::Pin,
     sync::Arc,
     task::{Context, Poll},
 };
-use tower::{Service, ServiceExt};
+use tokio::sync::oneshot;
+use tower::Service;
 
 use zebra_chain::block::{Block, BlockHeaderHash};
 use zebra_chain::types::BlockHeight;
 
-struct CheckpointVerifier<S> {
-    /// The underlying `ZebraState`.
-    state_service: S,
+/// The inner error type for CheckpointVerifier.
+///
+/// CheckpointVerifier returns two layers of `Result`s. The outer error type is
+/// fixed by the `tokio::sync::oneshot` channel receiver future.
```

Comment:
This info should probably be associated with the CheckpointVerifier and not the Error type def

Reply:
This info will be obsoleted when we remove the outer Result. See https://github.com/ZcashFoundation/zebra/pull/524#discussion_r447937237

I'll move any remaining info to the verifier.